PhpAmqpLib/Wire/IO/SocketIO.php

```diff
@@ -194,4 +219,40 @@ protected function enable_keepalive()
 
         socket_set_option($this->sock, SOL_SOCKET, SO_KEEPALIVE, 1);
     }
+
+    /**
+     * Heartbeat logic: check connection health here
+     */
+    protected function check_heartbeat()
+    {
+        // ignore unless heartbeat interval is set
+        if ($this->heartbeat !== 0 && $this->last_read && $this->last_write) {
+            $t = microtime(true);
+            $t_read = round($t - $this->last_read);
+            $t_write = round($t - $this->last_write);
+
+            // server has gone away
+            if (($this->heartbeat * 2) < $t_read) {
+                $this->reconnect();
+            }
+
+            // time for client to send a heartbeat
+            if (($this->heartbeat / 2) < $t_write) {
+                $this->write_heartbeat();
+            }
+        }
+    }
+
+    /**
+     * Sends a heartbeat message
+     */
+    protected function write_heartbeat()
+    {
+        $pkt = new AMQPWriter();
```

Comment:
Heartbeats are usually sent 5-20 times a minute (the interval is configurable). If that's too much GC overhead for you, you shouldn't be using a language such as PHP to begin with.